newton/utils/render.py

```diff
@@ -464,9 +506,193 @@ class SimRendererUsd(CreateSimRenderer(renderer=UsdRenderer)):
             renderer.end_frame()
             renderer.save()  # Save the USD file
     """
+    def __init__(
+        self,
+        model: newton.Model,
+        stage: str | Usd.Stage = None,
+        scaling: float = 1.0,
+        fps: int = 60,
+        up_axis: newton.AxisType | None = None,
+        show_rigid_contact_points: bool = False,
+        contact_points_radius: float = 1e-3,
+        show_joints: bool = False,
+        path_body_map: dict = None,
+        path_body_relative_transform: dict = None,
+        builder_results: dict = None,
+        **render_kwargs
+    ):
+        super().__init__(
+            model=model,
+            stage_or_title=stage,
+            scaling=scaling,
+            fps=fps,
+            up_axis=up_axis,
+            show_rigid_contact_points=show_rigid_contact_points,
+            contact_points_radius=contact_points_radius,
+            show_joints=show_joints,
+            **render_kwargs
+        )
+        self.path_body_map = path_body_map
+        self.path_body_map = path_body_map
+        self.path_body_relative_transform = path_body_relative_transform
+        self.builder_results = builder_results
+        self._prepare_output_stage()
+        self._precompute_parents_xform_inverses()
+
+    def render_time_sampled_transform(self, state: newton.State, sim_time):
+        """
+        Render transforms of USD prims.
+        """
+        time = sim_time * self.fps
+        body_q = state.body_q.numpy()
+        self.stage.SetEndTimeCode(round(time))
+        with Sdf.ChangeBlock():
+            for prim_path, body_id in self.path_body_map.items():
+                full_xform = body_q[body_id]
+                # TODO: do this once in __init__
+                # TODO: sanity check this with Eric Heiden
+                # Take relative xform into account
+                rel_xform = self.path_body_relative_transform.get(prim_path)
+                if rel_xform:
+                    full_xform = wp.mul(full_xform, rel_xform)
+
+                full_xform = self._apply_parents_inverse_xform(full_xform, prim_path)
+                self._update_usd_prim_xform(prim_path, full_xform, sim_time)
+
+
+    def render(
+        self,
+        state: newton.State,
+        sim_time,
+    ):
+        self.render_time_sampled_transform(state, sim_time)
```

Comment:
Does that mean the USD renderer now cannot render particles, cloths, tet meshes anymore? Only rigid bodies?

Reply:
We fix this comment by adding a new member function called `render_update_stage`.